scripts/deploy

```diff
@@ -0,0 +1,79 @@
+#!/bin/bash
+
+#                                   #
+# Deploys the Hack as a Service API #
+#                                   #
+
+set -e
+
+# Error handler
+trap 'echo -e "\\n-------------------- SOMETHING WENT WRONG. --------------------\\n\\nYou may need to remove ~/.haas_deploy_lock before trying again."' ERR
+
+# Check/set deploy lock
+ssh -T -i $KEY_PATH deploy@$HAAS_IP << EOF
+    if [[ ! -f ~/.haas_deploy_lock ]];
+    then
+        # Set up stuff
+        rm -rf ~/haas_api
+        touch ~/.haas_deploy_lock
+    else
+        echo 'Deploy already in progress'
+        exit 1
+    fi
+EOF
+
+# Upload files to server
+echo "Uploading source to server..."
+rsync -zr --exclude ".git/" --exclude "target/" -e "ssh -i $KEY_PATH" . deploy@$HAAS_IP:~/haas_api
+
+ssh -T -i $KEY_PATH deploy@$HAAS_IP << EOF
+    set -e
+
+    cd ~/haas_api
+
+    # Build some images
+    docker build -t haas_api -f docker/prod.Dockerfile .
+    docker build -t haas_migrator -f docker/migrator.Dockerfile .
+
+    # Migrate the DB schema
+    echo -e "\nMigrating database..."
+    docker run --rm -e "DATABASE_URL=$DATABASE_URL" --network haas_admin haas_migrator
+
+    echo -e "\nCreating haas_api_next..."
+    HAAS_API_NEXT=\$(docker run -d \
+        --restart always \
+        --env-file ~/.env \
+        --network haas_admin \
+        --name haas_api_next \
+        -v /var/run/docker.sock:/var/run/docker.sock \
+        haas_api)
+    
+    HAAS_API_NEXT_IP=\$(docker inspect \$HAAS_API_NEXT | jq -r .[0].NetworkSettings.Networks.haas_admin.IPAddress)
+
+    echo "New container IP: \$HAAS_API_NEXT_IP"
+    
+    # Wait 5 seconds for the API to start up
+    echo -e "\nWaiting impatiently for HaaS to start up..."
+    sleep 5
+
+    echo -e "\nSwitching reverse proxy over to new deployment..."
+    # Update Caddy config
+    curl -fs -X PATCH -H "Content-Type: application/json" -d "{\"@id\": \"haas_upstream\", \"dial\": \"\$HAAS_API_NEXT_IP:3000\"}" http://localhost:2019/id/haas_upstream
+
+    echo -e "\nRenaming containers..."
+    docker rename haas_api haas_api_prev || true
+    docker rename haas_api_next haas_api
+
+    echo -e "\nWaiting to make sure all is well with the reverse proxy..."
+    sleep 5
```

Comment:
Same as above - how long does it usually take for Caddy to settle after we make a config update?

Reply:
Again, it's just a paranoid safeguard (Caddy config updates are instant too)

Reply:
Got it, sounds good